File: Week03/98.验证二叉搜索树.py
#
# @lc app=leetcode.cn id=98 lang=python3
#
# [98] 验证二叉搜索树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        # 递归
        def helper(node, lower = float("-inf"), upper = float("inf")):
            if not node: return True
            
            if node.val <= lower or node.val >= upper:
                return False
            
            if not helper(node.left, lower, node.val):
                return False
            if not helper(node.right, node.val, upper):
                return False
            return True

        return helper(root)
# @lc code=end

File: Week03/98.验证二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=98 lang=cpp
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    long pre_min = LONG_MIN;
    bool isValidBST(TreeNode* root) {
        // 中序遍历
        if (root == nullptr) return true;
        if (false == isValidBST(root->left)){
            return false;
        }

        if (pre_min >= root->val) return false;

        pre_min = root->val;

        return isValidBST(root->right);    
    }
};
// @lc code=end

